### ISSUE TITLE
The method of organizing test data may affect the prediction results

### ISSUE THREAD

**Issue by ldz7** [jhagenauer/gwann#12]
Hello, jhagenauer, 

Recently I have done some analysis using this wonderful job. A problem I came across is that, when I organize data in test set in different ways, the prediction results may change. This really confused me. It can be illustrated with the example code in README.md. Below I give three cases to describle the problem in detail.

## Case One
```R
library(viridis)
library(gwann)
library(ggplot2)

set.seed(1)
data(toy4)

x<-as.matrix(toy4[,c("x1","x2")])
y<-as.numeric(toy4[,c("y")] )
dm<-as.matrix(dist(toy4[,c("lon","lat")])  )
s_test<-sample(nrow(x),0.3*nrow(x)) # indices of test samples

x_pred1 <- x[s_test,]
w_pred1 <- dm[-s_test, s_test]
r<-gwann(x_train=x[-s_test,],y_train=y[-s_test],w_train=dm[-s_test,-s_test],
         x_pred=x_pred1,w_pred=w_pred1,
         nrHidden=30,batchSize=50,lr=0.1,
         adaptive=F,
         bwSearch="goldenSection",bwMin=min(dm)/4, bwMax=max(dm)/4,
         threads=8
)
p<-diag(r$predictions)
print(p[1:5])
```
This is exactly the example code in the readme file, except that I have added a seed to make the result comparable. The predition results, i.e., the varaible `p` in the code is of my interest. 

The output is 
```
Golden section search...
Cross-validation results for hyperparameter search (folds: 10, repeats: 1):
* Bandwidth: 2.0720938004226177
* Iterations: 1606
* RMSE: 0.3637624585267835
Building final model with bandwidth 2.0720938004226177 and 1606 iterations...
[1] 4.6746140 8.2461332 0.6771165 0.2575546 2.5992905
```

## Case Two
```R
library(viridis)
library(gwann)
library(ggplot2)

set.seed(1)
data(toy4)

x<-as.matrix(toy4[,c("x1","x2")])
y<-as.numeric(toy4[,c("y")] )
dm<-as.matrix(dist(toy4[,c("lon","lat")])  )
s_test<-sample(nrow(x),0.3*nrow(x)) # indices of test samples

x_pred2 <- x[s_test[1:5],]
w_pred2 <- dm[-s_test, s_test[1:5]]
r<-gwann(x_train=x[-s_test,],y_train=y[-s_test],w_train=dm[-s_test,-s_test],
         x_pred=x_pred2,w_pred=w_pred2,
         nrHidden=30,batchSize=50,lr=0.1,
         adaptive=F,
         bwSearch="goldenSection",
         bwMin=min(dm)/4, bwMax=max(dm)/4,
         threads=8
)
p<-diag(r$predictions)
print(p[1:5])
```
```
Golden section search...
Cross-validation results for hyperparameter search (folds: 10, repeats: 1):
* Bandwidth: 2.0720938004226177
* Iterations: 1606
* RMSE: 0.3637624585267835
Building final model with bandwidth 2.0720938004226177 and 1606 iterations...
[1] 4.6890841 8.4763250 0.4011214 0.4280682 2.5759752
```
Here I only select the first five observations of test data in Case One to form a new test set `x_pred2` and corresponding weight `w_pred2`. From the output we can see that although the cross-validation results are the same, the prediction results are quite different from those in Case One (especially for the third and fourth observations), which is very strange. 

## Case Three
```R
library(viridis)
library(gwann)
library(ggplot2)

set.seed(1)
data(toy4)

x<-as.matrix(toy4[,c("x1","x2")])
y<-as.numeric(toy4[,c("y")] )
dm<-as.matrix(dist(toy4[,c("lon","lat")])  )
s_test<-sample(nrow(x),0.3*nrow(x)) # indices of test samples

s_test3 <- s_test
s_test3[2] <- s_test3[1]
x_pred3 <- x[s_test3,]
w_pred3 <- dm[-s_test, s_test3]
r<-gwann(x_train=x[-s_test,],y_train=y[-s_test],w_train=dm[-s_test,-s_test],
         x_pred=x_pred3,w_pred=w_pred3,
         nrHidden=30,batchSize=50,lr=0.1,
         adaptive=F,
         bwSearch="goldenSection",
         bwMin=min(dm)/4, bwMax=max(dm)/4,
         threads=8
)
p<-diag(r$predictions)
print(p[1:5])
```
```
Golden section search...
Cross-validation results for hyperparameter search (folds: 10, repeats: 1):
* Bandwidth: 2.0720938004226177
* Iterations: 1606
* RMSE: 0.3637624585267835
Building final model with bandwidth 2.0720938004226177 and 1606 iterations...
[1] 4.6732400 4.6189392 0.6733432 0.2431097 2.5998484
```
In this case, I replace the second observation of test set in Case One with the first observation. So the first two observations are totally the same. And I expect the predicted values of these two observations to be the same (because their data and distance vector are the same). However, the ouput shows that they are different.

***
In summary, there are at least two strange things happened in the above three cases:
1. Using subset of a test data will give a different prediction (Case One vs. Case Two);
2. Two identical observations in the test set will result in different prediction results (Case Three).

Did I do something wrong? Hope for your help!

**Comment by SoviaIndrawati** [comment]
Hello, Mr. Julian.
I tried to install gwann with this command in Windows 10 :
Sys.setenv("R_REMOTES_NO_ERRORS_FROM_WARNINGS" = "true")
> if (!require("devtools"))
+     install.packages("devtools")
> devtools::install_github("jhagenauer/gwann",INSTALL_opts=c("--no-multiarch"))

And i got this error:
Downloading GitHub repo jhagenauer/gwann@HEAD
Error in utils::download.file(url, path, method = method, quiet = quiet,  : 
  download from 'https://api.github.com/repos/jhagenauer/gwann/tarball/HEAD' failed
Warning message:
In devtools::install_github("jhagenauer/gwann", INSTALL_opts = c("--no-multiarch")) :
  Arguments in `...` must be used.
✖ Problematic argument:
• INSTALL_opts = c("--no-multiarch")
ℹ Did you misspell an argument name?

Do you have any suggestion how to solve the problem?
Thank you so much in advance. May you have a blessed day.
